module/zfs/arc.c

```diff
@@ -8958,104 +9025,280 @@ EXPORT_SYMBOL(arc_getbuf_func);
 EXPORT_SYMBOL(arc_add_prune_callback);
 EXPORT_SYMBOL(arc_remove_prune_callback);
 
-/* BEGIN CSTYLED */
-module_param(zfs_arc_min, ulong, 0644);
-MODULE_PARM_DESC(zfs_arc_min, "Min arc size");
-
-module_param(zfs_arc_max, ulong, 0644);
-MODULE_PARM_DESC(zfs_arc_max, "Max arc size");
+static int
+param_set_arc_tunable(const char *val, zfs_kernel_param_t *kp)
+{
+	int ret;
 
-module_param(zfs_arc_meta_limit, ulong, 0644);
-MODULE_PARM_DESC(zfs_arc_meta_limit, "Meta limit for arc size");
+	ret = param_set_ulong(val, kp);
+	if (ret < 0)
+		return (ret);
 
-module_param(zfs_arc_meta_limit_percent, ulong, 0644);
-MODULE_PARM_DESC(zfs_arc_meta_limit_percent,
-	"Percent of arc size for arc meta limit");
+	arc_tuning_update();
 
-module_param(zfs_arc_meta_min, ulong, 0644);
-MODULE_PARM_DESC(zfs_arc_meta_min, "Min arc metadata");
+	return (ret);
+}
 
-module_param(zfs_arc_meta_prune, int, 0644);
-MODULE_PARM_DESC(zfs_arc_meta_prune, "Meta objects to scan for prune");
+/*
+ * The passed string value is expected to include one of the allowed states
+ * listed below and an allowed type.
+ *
+ * Allowed States:
+ * - mru       - Only flush the MRU list
+ * - mfu       - Only flush the MFU list
+ * - mru_ghost - Only flush the MRU ghost list
+ * - mfu_ghost - Only flush the MFU ghost list
+ *
+ * Allowed Types:
+ * - data      - Only flush data
+ * - metadata  - Only flush metadata
+ *
+ * Aliases:
+ * - all-states = mru, mfu, mru_ghost, mfu_ghost
+ * - all-types  = data, metadata
+ * - all        = all-states, all-types
+ */
+enum {
+	TOKEN_STATE_MRU,
+	TOKEN_STATE_MFU,
+	TOKEN_STATE_MRU_GHOST,
+	TOKEN_STATE_MFU_GHOST,
+	TOKEN_STATE_L2ARC,
+	TOKEN_STATE_ALL,
+	TOKEN_TYPE_DATA,
+	TOKEN_TYPE_METADATA,
+	TOKEN_TYPE_ALL,
+	TOKEN_POOL_NAME,
+	TOKEN_ALL,
+	TOKEN_LAST,
+};
 
-module_param(zfs_arc_meta_adjust_restarts, int, 0644);
-MODULE_PARM_DESC(zfs_arc_meta_adjust_restarts,
-	"Limit number of restarts in arc_adjust_meta");
+static const match_table_t flush_tokens = {
+	{ TOKEN_STATE_MRU,		"mru" },
+	{ TOKEN_STATE_MFU,		"mfu" },
+	{ TOKEN_STATE_MRU_GHOST,	"mru_ghost" },
+	{ TOKEN_STATE_MFU_GHOST,	"mfu_ghost" },
+	{ TOKEN_STATE_L2ARC,		"l2arc" },
+	{ TOKEN_STATE_ALL,		"all-states" },
+	{ TOKEN_TYPE_DATA,		"data" },
+	{ TOKEN_TYPE_METADATA,		"metadata" },
+	{ TOKEN_TYPE_ALL,		"all-types" },
+	{ TOKEN_POOL_NAME,		"pool=%s" },
+	{ TOKEN_ALL,			"all" },
+	{ TOKEN_LAST,			NULL },
+};
 
-module_param(zfs_arc_meta_strategy, int, 0644);
-MODULE_PARM_DESC(zfs_arc_meta_strategy, "Meta reclaim strategy");
+static int
+parse_arc_flush(char *option, int token, substring_t *args, flush_state_t *fs)
```

Comment:
Doing this kind of parsing in the kernel really isn't ideal, although it was expedient at the time.  We should pull this up in to `zinject(8)` and pass the resulting flags in the `zinject_record_t`.  We're only talking about 8 bits and we could claim all 32 from the `zi_pad`.